File: android_mobile/src/main/java/com/alexstyl/specialdates/util/HashMapList.java
package com.alexstyl.specialdates.util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

public class HashMapList<K, V> {

    private final HashMap<K, List<V>> map = new HashMap<>();

    public boolean addValue(K key, V value) {
        List<V> listValue = map.get(key);
        if (listValue == null) {
            listValue = new ArrayList<>();
            map.put(key, listValue);
        }
        return listValue.add(value);
    }

    public List<V> get(K key) {
        return map.get(key);
    }

    public boolean isEmpty() {
        return map.isEmpty();
    }

    public Set<K> keys() {
        return map.keySet();
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/util/HashMapList.kt
package com.alexstyl.specialdates.util

class HashMapList<K, V> {

    private val map = HashMap<K, MutableList<V>>()

    fun addValue(key: K, value: V): Boolean {
        if (!map.containsKey(key)) {
            map.put(key, ArrayList())
        }
        return map[key]!!.add(value)
    }

    operator fun get(key: K): List<V>? = map[key]

    val isEmpty: Boolean
        get() = map.isEmpty()

    fun keys(): Set<K> = map.keys

    fun clear() {
        map.clear()
    }
}
